File: core/app/models/beeper/signal.rb
Beeper::Signal = Data.define(:status, :title, :message, :metrics) do
  def initialize(status:, title: nil, message: nil, metrics: {})
    status_sym = status.to_sym
    unless [ :ok, :alerting, :error ].include?(status_sym)
      raise ArgumentError, "status must be :ok, :alerting, or :error (got #{status.inspect})"
    end

    super(status: status_sym, title: title, message: message, metrics: metrics || {})
  end

  def ok?
    status == :ok || status == "ok"
  end

  def alerting?
    status == :alerting || status == "alerting"
  end

  def error?
    status == :error || status == "error"
  end

  def to_h
    {
      "status" => status.to_s,
      "title" => title,
      "message" => message,
      "metrics" => metrics
    }.compact
  end
end

File: apps/cli/internal/probe/signal.go
package probe

type Status string

const (
	StatusOk       Status = "ok"
	StatusAlerting Status = "alerting"
	StatusError    Status = "error"
)

type Signal struct {
	Status  Status         `json:"status"`
	Title   string         `json:"title"`
	Message string         `json:"message,omitempty"`
	Metrics map[string]any `json:"metrics,omitempty"`
}

func OkSignal(title string, message string, metrics map[string]any) *Signal {
	return &Signal{
		Status:  StatusOk,
		Title:   title,
		Message: message,
		Metrics: metrics,
	}
}

func AlertingSignal(title string, message string, metrics map[string]any) *Signal {
	return &Signal{
		Status:  StatusAlerting,
		Title:   title,
		Message: message,
		Metrics: metrics,
	}
}

func ErrorSignal(title string, message string, metrics map[string]any) *Signal {
	return &Signal{
		Status:  StatusError,
		Title:   title,
		Message: message,
		Metrics: metrics,
	}
}
